Add argon/classic osu!mania combo counter

frenzibyte opened the pull request:
- Closes https://github.com/ppy/osu/issues/20541
- Supersedes #20579 
- [x] Depends on https://github.com/ppy/osu/pull/26249

Adds support for classic osu!mania combo counter, as well as a very basic implementation for "Argon" skin to increase its usability (cc @peppy for adding animations to the argon one if wanted).

This does not include turning the combo colour to yellow during a hold note, especially since hold notes don't have ticks anymore, which makes the yellow state look odd.

Preview:

https://github.com/ppy/osu/assets/22781491/fae36613-f0e8-4bed-8c1c-81ee7610919f

Argon preview (again, not final):

https://github.com/ppy/osu/assets/22781491/2ad44d09-3ec2-4714-8d73-ff2f345b2ce7

frenzibyte committed:
Disable "closest" anchor in mania combo counter for convenience

frenzibyte committed:
Add legacy mania combo counter lookups

frenzibyte committed:
Add legacy mania combo counter implementation

frenzibyte committed:
Add very basic argon mania combo counter implementation

frenzibyte committed:
Add various visual test coverage

frenzibyte committed:
Abstractify `LegacyComboCounter` to re-use for mania

frenzibyte committed:
Fix failing tests

mcendu commented:
I think this can be used for taiko as well?

frenzibyte commented:
Yes but it's not in scope of this PR.

peppy committed:
Merge branch 'master' into mania-combo-counter

peppy committed:
Update in line with new changes

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -236,6 +237,9 @@ private void load(OsuConfigManager config, OsuGameBase game, CancellationToken c
             DrawableRuleset = ruleset.CreateDrawableRulesetWith(playableBeatmap, gameplayMods);
             dependencies.CacheAs(DrawableRuleset);
 
+            if (DrawableRuleset is IDrawableScrollingRuleset scrollingRuleset)
```
This is a bit ad-hoc isn't it

frenzibyte replied:
I...probably can just resolve the ruleset and get the scrolling info from there.

peppy replied:
Yeah not sure. Maybe it's fine. Just a bit hard to track what DI dependencies are available right now.

peppy replied:
I'm gonna leave this as-is for now. Rationale is that `IScrollingInfo` is used in a lot of places within `DrawableRuleset` so I'd expect for it to also be available here.

peppy commented on osu.Game/Skinning/ISerialisableDrawable.cs:
```diff
@@ -27,6 +27,14 @@ public interface ISerialisableDrawable : IDrawable
+        /// Whether this component supports the "closest" anchor.
+        /// </summary>
+        /// <remarks>
+        /// This is disabled by some components that shift position automatically.
```
What's an example of this? I don't quite understand.

frenzibyte replied:
Disabled by `LegacyManiaComboCounter` because its position is shifted when the user flips the scrolling direction. I don't recall why though.

peppy replied:
This seems like an odd choice. So the user can still choose an anchor and it will just flip? I'd say that closest should still be allowed and the flip is only applied if the counter is in a certain region of the screen (or have a certain anchor applied).

peppy replied:
I've removed this for now as I can't really understand why it is required. It would make more sense to completely disable anchors (or a certain subset) but honestly I'd just leave this up to the user, and adjust the automatic flipping logic to only work in the cases where it makes sense.

peppy commented on osu.Game/Skinning/LegacyDefaultComboCounter.cs:
```diff
@@ -0,0 +1,85 @@
+    /// <summary>
+    /// Uses the 'x' symbol and has a pop-out effect while rolling over.
+    /// </summary>
+    public partial class LegacyDefaultComboCounter : LegacyComboCounter
```
I'm honestly not sure this abstraction was required. Could have just copy-pasted and spent less effort? And potentially had more maintainable / legible code.

frenzibyte replied:
Sure.

peppy replied:
No need to change it at this point, I'll do so if the review is painful enough.

peppy committed:
Remove `SupportsClosestAnchor` for the time being

This may have had a good reason to be added, but I can't find that
reason, so let's keep things simple for the time being.

peppy committed:
Remove unnecessary `UsesFixedAnchor` specifications

peppy committed:
Simplify argon mania combo counter implementation by sharing with base counter

peppy commented on osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:
```diff
@@ -0,0 +1,42 @@
+            direction = scrollingInfo.Direction.GetBoundCopy();
+            direction.BindValueChanged(_ => updateAnchor());
+
+            // two schedules are required so that updateAnchor is executed in the next frame,
+            // which is when the combo counter receives its Y position by the default layout in ArgonManiaSkinTransformer.
```
really dunno but willing to overlook for now because it works.

peppy commented:
@frenzibyte I've made sizable changes to this PR. Want to give it a once-over?

peppy committed:
Fix migration failure due to change in class name

frenzibyte commented on osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs:
```diff
@@ -45,6 +45,9 @@ public OsuLegacySkinTransformer(ISkin skin)
             switch (lookup)
             {
                 case SkinComponentsContainerLookup containerLookup:
+                    if (containerLookup.Target != SkinComponentsContainerLookup.TargetArea.MainHUDComponents)
+                        return base.GetDrawableComponent(lookup);
+
```
Nit-pick but I don't expect this to be placed here:
```diff
diff --git a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
index df8c04638d..2ba87eda2c 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
@@ -31,9 +31,6 @@ public CatchLegacySkinTransformer(ISkin skin)
             switch (lookup)
             {
                 case SkinComponentsContainerLookup containerLookup:
-                    if (containerLookup.Target != SkinComponentsContainerLookup.TargetArea.MainHUDComponents)
-                        return base.GetDrawableComponent(lookup);
-
                     // Modifications for global components.
                     if (containerLookup.Ruleset == null)
                         return base.GetDrawableComponent(lookup);
@@ -43,27 +40,33 @@ public CatchLegacySkinTransformer(ISkin skin)
                         return d;
 
                     // Our own ruleset components default.
-                    // todo: remove CatchSkinComponents.CatchComboCounter and refactor LegacyCatchComboCounter to be added here instead.
-                    return new DefaultSkinComponentsContainer(container =>
+                    switch (containerLookup.Target)
                     {
-                        var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
-
-                        if (keyCounter != null)
-                        {
-                            // set the anchor to top right so that it won't squash to the return button to the top
-                            keyCounter.Anchor = Anchor.CentreRight;
-                            keyCounter.Origin = Anchor.CentreRight;
-                            keyCounter.X = 0;
-                            // 340px is the default height inherit from stable
-                            keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
-                        }
-                    })
-                    {
-                        Children = new Drawable[]
-                        {
-                            new LegacyKeyCounterDisplay(),
-                        }
-                    };
+                        case SkinComponentsContainerLookup.TargetArea.MainHUDComponents:
+                            // todo: remove CatchSkinComponents.CatchComboCounter and refactor LegacyCatchComboCounter to be added here instead.
+                            return new DefaultSkinComponentsContainer(container =>
+                            {
+                                var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
+
+                                if (keyCounter != null)
+                                {
+                                    // set the anchor to top right so that it won't squash to the return button to the top
+                                    keyCounter.Anchor = Anchor.CentreRight;
+                                    keyCounter.Origin = Anchor.CentreRight;
+                                    keyCounter.X = 0;
+                                    // 340px is the default height inherit from stable
+                                    keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
+                                }
+                            })
+                            {
+                                Children = new Drawable[]
+                                {
+                                    new LegacyKeyCounterDisplay(),
+                                }
+                            };
+                    }
+
+                    return null;
 
                 case CatchSkinComponentLookup catchSkinComponent:
                     switch (catchSkinComponent.Component)
```

I also don't want to imagine that there is any other class that will return ruleset-specific components for the legacy skin other than this one, so I'm returning null at the end of the diff above.

Maybe this is irrelevant anyway because these methods will be refactored later on.

frenzibyte replied:
Also why not put the "Skin has configuration" if statement to the top of the method?

peppy replied:
I've applied something similar to your diff. Still waiting on feedback from you for the issues you encountered during testing.

frenzibyte reviewed (commented):
Combo counter does not seem to show up on argon skin:
![CleanShot 2024-08-10 at 11 20 20](https://github.com/user-attachments/assets/b326a4e4-0f04-42f8-9bbd-e382e4bd99e5)

I also kinda don't like the "x" symbol in the argon counter, looks weird when it's placed on top of the playfield. Would try to remove it.

It's also incorrectly pushed away on classic skin:
![CleanShot 2024-08-10 at 11 19 44](https://github.com/user-attachments/assets/ed89f928-5fb9-4dd9-99e9-926eb67d4ff1)

peppy commented:
> Combo counter does not seem to show up on argon skin

can you look into this? it shows fine here.

peppy committed:
Standardise skin transformer code structure

frenzibyte commented:
I began investigating this and posted some [observations in discord](https://discord.com/channels/188630481301012481/188630652340404224/1272481231245807711) until all of a sudden everything started working again and I have no idea why.

frenzibyte commented on osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:
```diff
@@ -0,0 +1,42 @@
+
+        private void updateAnchor()
+        {
+            Anchor &= ~(Anchor.y0 | Anchor.y2);
+            Anchor |= direction.Value == ScrollingDirection.Up ? Anchor.y2 : Anchor.y0;
```
The legacy part handles the case where the anchor is set to center, but this one doesn't?

peppy replied:
I didn't change this 🙈

frenzibyte replied:
Whoops 😅

peppy replied:
I've updated this to match but there still seems to be something weird going on here.

- Set scroll direction to "Down"
- Put the combo counter near the bottom of the screen
- Change scroll direction

The combo won't move the first time changing direction.

frenzibyte replied:
Alleviated the issue in https://github.com/ppy/osu/pull/26254/commits/66adddbfb8020c019bbdf672ade3f00b4aead7ef. This is the reason why "Closest" anchor should not be used.

To explain briefly, I am using the sign of the Y coordinate as an indicator for whether the combo counter is in the correct position or not. When the scroll direction is "Down", the anchor is set to top and the Y position is made sure to be positive. When the scroll direction is "Up", the anchor is set to bottom and the Y position is made sure to be negative.

When the "Closest" anchor is selected, if the user moves their combo counter from top to bottom or vice versa, the anchor will change, and defies everything that the logic above does, therefore breaking apart.

frenzibyte commented on osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace osu.Game.Rulesets.Mania.Skinning.Argon
+{
+    public partial class ArgonManiaComboCounter : ArgonComboCounter
```
As I mentioned already I think the "x" symbol next to the counter feels weird on mania.

frenzibyte commented on osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs:
```diff
@@ -26,6 +28,34 @@ public ManiaArgonSkinTransformer(ISkin skin, IBeatmap beatmap)
             switch (lookup)
             {
+                case SkinComponentsContainerLookup containerLookup:
+                    if (containerLookup.Target != SkinComponentsContainerLookup.TargetArea.MainHUDComponents)
+                        return base.GetDrawableComponent(lookup);
+
+                    // Only handle per ruleset defaults here.
+                    if (containerLookup.Ruleset == null)
+                        return base.GetDrawableComponent(lookup);
+
+                    // Skin has configuration.
+                    if (base.GetDrawableComponent(lookup) is UserConfiguredLayoutContainer d)
+                        return d;
```
Both this and `ManiaLegacySkinTransformer` haven't been updated in line with d2eb6ccb8caa

peppy replied:
Have updated these.

frenzibyte commented:
As you touched in [discord](https://discord.com/channels/188630481301012481/188630652340404224/1272489428669169725), it turns out that the cause of this issue is that I had "beatmap skins" enabled, which makes the beatmap skin layer return a `LegacyComboCounter` with a zero combo position.

The reason why the combo position is zero is because of the lookup only happening on the beatmap skin layer, and the beatmap skin layer disables mania lookups entirely (see `AllowManiaConfigLookups` override), therefore the fallback value in the lookup line (which is `0`) is assigned and therefore the combo counter is stuck at the top.

I'm not sure what is the best way to solve this, one way could be to use the `isLegacySkin` flag which determines whether the skin is user or beatmap and then return null when returning the default mania HUD layout if it's a beatmap skin, but I'm not really sure that's a good way to handle things.

peppy committed:
Fix file access test failure by forcing retries

See https://github.com/ppy/osu/actions/runs/10369630825/job/28708248682.

peppy committed:
Merge branch 'master' into mania-combo-counter

peppy committed:
Fix incorrect anchor handling in `ArgonManiaComboCounter`

frenzibyte committed:
Remove position-flipping logic from mania combo counters for now

We need a general method to do this amicably, such as an HUD target that flips the position of its children when the direction is flipped. Something to consider later.

frenzibyte committed:
Remove `x` symbol from argon mania combo counter

frenzibyte committed:
Actually bring back position-flipping logic and disable "closest" anchor

frenzibyte committed:
Fix beatmap skin on mania breaking HUD apart

frenzibyte committed:
Update code order to match everything else

frenzibyte committed:
Try make code look better
